vyxal/lexer.py

```diff
@@ -0,0 +1,193 @@
+        Each token is represented as a Token object.
+    """
+
+    tokens: list[Token] = []
+    source: collections.deque = collections.deque(source)
+
+    contextual_token_value: str = ""
+
+    while source:
+        # By treating the program as a queue, we can dequeue elements
+        # until a certain predicate is satisfied. In simple terms, this
+        # means it's easier to group things based on order...you don't
+        # have to worry about what you group first.
+
+        head: str = source.popleft()
+        if head == "\\":  # Need to escape the next character
+            if source:
+                # This has the consequence of making backslahses at the
+                # end of a program not error.
+                tokens.append(Token(TokenType.LITERAL, source.popleft()))
+
+        elif head in "`»«":  # String
+            # Dequeue characters until the same string character is
+            # reached.
+            contextual_token_value = ""
+            while source and source[0] != head:
+                character: str = source.popleft()
+                if head == "`" and character == "\\":
+                    # Handle the escape by just dequeueing the next
+                    # character
+                    if source:
+                        contextual_token_value += "\\" + source.popleft()
+                else:
+                    contextual_token_value += character
+            token_type: str = ""
+            if head == "`":
+                token_type = TokenType.LITERAL
+            elif head == "»":
+                token_type = TokenType.COMPRESSED_NUMBER
+            elif head == "«":
+                token_type = TokenType.COMPRESSED_STRING
+            tokens.append(Token(token_type, contextual_token_value))
+            if source:
+                source.popleft()
+        elif head in string.digits + ".":
+            contextual_token_value = head
+            while source and source[0] in string.digits + ".":
+                contextual_token_value += source.popleft()
+            tokens.append(Token(TokenType.LITERAL, contextual_token_value))
+        elif head == "‛":
+            contextual_token_value = ""
+            while source and len(contextual_token_value) != 2:
+                contextual_token_value += source.popleft()
+            tokens.append(Token(TokenType.LITERAL, contextual_token_value))
+        elif head in "@→←°":
+            tokens.append(Token(TokenType.GENERAL, head))
+            contextual_token_value = ""
+            while source and source[0] in string.ascii_letters + "_":
+                contextual_token_value += source.popleft()
+
+            tokens.append(Token(TokenType.NAME, contextual_token_value))
+        elif head in "k∆øÞ¨":
+            if source:
+                tokens.append(Token(TokenType.GENERAL, head + source.popleft()))
+        else:
+            tokens.append(Token(TokenType.GENERAL, head))
+    return tokens
+
+
+if __name__ == "__main__":
```

Comment:
I think these should be run as proper tests (compare against an expected output and throw an error if it doesn't work). It doesn't feel right to be running this module all by itself, and IMO the only module with an `if __name__ == "__main__"` should be the interpreter/compiler.

Reply:
What do mean by proper tests?

Reply:
Well, each group of tests would be in a function. I was thinking all the tests for the lexer would be in a `lexer_tests.py` file with all the other tests. That way, if something fails, we'll know when we push.

Reply:
who are you, actually?

Reply:
A username?

Reply:
I don't understand. In real life? I'm not telling you that, for obvious reasons.